core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +669,824 @@ impl CheckedAdd for Fixed64 {
+				other.lpad(self.len())
+			} else if other.len() > self.len() {
+				self.lpad(other.len())
+			}
+		}
+
+		/// Adds `self` with `other`. self and other do not have to have any particular size. Given
+		/// that the `n = max{size(self), size(other)}`, it will produce a number with `n + 1`
+		/// limbs.
+		///
+		/// This function does not strip the output and returns the original allocated `n + 1`
+		/// limbs. The caller may strip the output if desired.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn add(mut self, mut other: Self) -> Self {
+			Self::resize(&mut self, &mut other);
+			let n = self.len();
+			let mut k: Double = 0;
+			let mut w = Self::with_capacity(n + 1);
+
+			for j in 0..n {
+				let u = Double::from(self.get(j));
+				let v = Double::from(other.get(j));
+				let s = u + v + k;
+				w.set(j, (s % B) as Single);
+				k = s / B;
+			}
+			// k is always 0 or 1.
+			w.set(n, k as Single);
+			w
+		}
+
+		/// Subtracts `other` from `self`. self and other do not have to have any particular size.
+		/// Given that the `n = max{size(self), size(other)}`, it will produce a number of size `n`.
+		///
+		/// If `other` is bigger than `self`, `Err(B - borrow)` is returned.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn sub(mut self, mut other: Self) -> Result<Self, Self> {
+			Self::resize(&mut self, &mut other);
+			let n = self.len();
+			let mut k = 0;
+			let mut w = Self::with_capacity(n);
+			for j in 0..n {
+				let s = {
+					let u = Double::from(self.get(j));
+					let v = Double::from(other.get(j));
+					let mut needs_borrow = false;
+					let mut t = 0;
+
+					if let Some(v) = u.checked_sub(v) {
+						if let Some(v2) = v.checked_sub(k) {
+							t = v2 % B;
+							k = 0;
+						} else {
+							needs_borrow = true;
+						}
+					} else {
+						needs_borrow = true;
+					}
+					if needs_borrow {
+						t = u + B - v - k;
+						k = 1;
+					}
+					t
+				};
+				// PROOF: t either comes from `v2 % B`, or from `u + B - v - k`. The former is
+				// trivial. The latter will not overflow this branch will only happen if the sum of
+				// `u - v - k` part has been negative, hence `u + B - v - k < b`.
+				w.set(j, s as Single);
+			}
+
+			if k.is_zero() {
+				Ok(w)
+			} else {
+				Err(w)
+			}
+		}
+
+		/// Multiplies n-limb number `self` with m-limb number `other`.
+		///
+		/// The resulting number will always have `n + m` limbs.
+		///
+		/// This function does not strip the output and returns the original allocated `n + m`
+		/// limbs. The caller may strip the output if desired.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn mul(self, other: Self) -> Self {
+			let n = self.len();
+			let m = other.len();
+			let mut w = Self::with_capacity(m + n);
+
+			for j in 0..n {
+				if self.get(j) == 0 {
+					w.set(j + m, 0);
```

Comment:
Isn't this already 0?

Reply:
it for sure is. I was pedantically just following the written steps from the book 😁 👍